crates/solver/src/settlement_post_processing/optimize_unwrapping.rs

```diff
@@ -0,0 +1,206 @@
+use crate::settlement::Settlement;
+use anyhow::Result;
+use contracts::WETH9;
+use primitive_types::U256;
+
+/// Tries to do one of 2 optimizations.
+/// 1) Drop WETH unwraps and instead pay ETH with the settlment contract's buffer.
+/// 2) Top up settlement contract's ETH buffer by unwrapping way more WETH than this settlement
+///    needs. This will cause the next few settlements to use optimization 1.
+/// If this function returns Err(), the original settlement will not be modified in any way.
+pub async fn optimize_unwrapping<V, VFut, B, BFut>(
+    settlement: &mut Settlement,
+    settlement_would_succeed: V,
+    get_weth_balance: B,
+    weth: &WETH9,
+    unwrap_factor: f64,
+) -> Result<()>
+where
+    V: Fn(Settlement) -> VFut,
+    VFut: futures::Future<Output = bool>,
+    B: Fn() -> BFut,
+    BFut: futures::Future<Output = Result<U256>>,
+{
+    let required_eth_payout = settlement.encoder.amount_to_unwrap(weth.address());
+    if required_eth_payout.is_zero() {
+        return Ok(());
+    }
+
+    // simulate settlement without unwrap
+    let mut optimized_settlement = settlement.clone();
+    optimized_settlement.encoder.drop_unwrap(weth.address());
+
+    if settlement_would_succeed(optimized_settlement.clone()).await {
+        tracing::debug!("use internal buffer to unwraps");
+        *settlement = optimized_settlement;
+        return Ok(());
+    }
```

Comment:
At first, I thought that we could just use the condition `if required_eth_payout <= weth_balance { ... }` here (in order to avoid a node round-trip), but it turns out, you can't because it doesn't capture solvers using `ETH` buffers for some other purpose which could leave not enough `ETH` for the payout at the end. Maybe add a comment here as to why we need a simulation instead of just checking amounts?

Reply:
Done.